lib/ferto/client.rb

```diff
@@ -100,7 +101,15 @@ def download(aggr_id:, aggr_limit: @aggr_limit, url:,
           handle.connect_timeout = connect_timeout
           handle.timeout = timeout
         end
-      rescue Curl::Err::ConnectionFailedError => e
+
+        case res.response_code
+        when 400..599
+          error_msg = ("An error occured during the download call. "  \
+            "Received a #{res.response_code} response code and body " \
+            "#{res.body_str}")
+          raise Ferto::ResponseError.new(error_msg, res)
+        end
+      rescue Curl::Err::CurlError => e
```

Comment:
You are changing the `ConnectionFailedError` to `CurlError`here, is that intentional?

Reply:
Yes from now on we respond to all descendants of `Curl::Err::CurlError` with our custom `Ferto::ConnectionError`. So we can have a single exception handling code path.

Reply:
I am not a big fan of a "catch all" logic around the exception handling. I would suggest to retain the `ConnectionFailedError` and changing it to a `Ferto::ConnectionError` for the appropriate handling in our library but leaving the rest of the exceptions unhandled. It wouldn't be semantically consistent to change a curl generic error to ConnectionError just to handle it, would it?

Reply:
Right, so i will leave any curl errors unhandled and revert the logic in order for `Ferto::ConnectionError` to be raised only in case a `Curl::Err::ConnectionFailedError` error
occurs!

Reply:
Perfect! Thanks :+1: